packages/storage/amplify_storage_s3_dart/lib/src/prefix_resolver/storage_access_level_aware_prefix_resolver.dart

```diff
@@ -0,0 +1,60 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:amplify_storage_s3_dart/amplify_storage_s3_dart.dart';
+import 'package:meta/meta.dart';
+
+/// {@template amplify_storage_s3_dart.storage_access_level_aware_prefix_resolver}
+/// The default Storage S3 plugin prefix resolver that resolves prefix according
+/// to given [StorageAccessLevel] and currently signed user.
+/// E.g.
+/// * [StorageAccessLevel.guest] resolves `'public'`
+/// * [StorageAccessLevel.protected] resolves `'protected/<userIdentityId>'`
+/// * [StorageAccessLevel.private] resolves `'private/<userIdentityId>'`
+/// {@endtemplate}
+class StorageAccessLevelAwarePrefixResolver
+    implements AmplifyStorageS3PrefixResolver {
+  /// {@macro amplify_storage_s3_dart.storage_access_level_aware_prefix_resolver}
+  StorageAccessLevelAwarePrefixResolver({
+    String? delimiter,
+  }) : _delimiter = delimiter ?? '/';
+
+  late final TokenAmplifyAuthProvider _identityProvider;
```

Comment:
Can this be injected?

Reply:
You meant to discuss it "should be"/"should not be" injected?

The auth provider becomes available only when `plugin.configure` is being called, so it comes `late`, where I inject it into the default prefix resolver as a dependency. I'm open for suggestions.

Reply:
I would find a way to make it a `final` parameter injected via the constructor

Reply:
Hum yea, can initiate the prefix resolver when `configure` is called where can make the auth provider a `final` inside the prefix resolver.